File: java/bench/Bench.java
// Performance bench for the Java port. Emits one JSON line per
// build/bench/README.md; diagnostics go to stderr. Uses the default-package
// Struct compiled alongside this file.
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

import voxgig.struct.Struct;

public class Bench {
    static long sink = 0;

    interface Fn { void run(); }

    static int envi(String k, int d) {
        try { return Integer.parseInt(System.getenv(k)); }
        catch (Exception e) { return d; }
    }

    static Object build(int w, int d, int leaf) {
        if (d == 0) return leaf;
        HashMap<String, Object> m = new HashMap<>();
        for (int i = 0; i < w; i++) m.put("k" + i, build(w, d - 1, leaf));
        return m;
    }

    static long nodecount(int w, int d) {
        long n = 0, p = 1;
        for (int i = 0; i <= d; i++) { n += p; p *= w; }
        return n;
    }

    static double[] measure(int warm, int runs, Fn f) {
        for (int i = 0; i < warm; i++) f.run();
        double[] t = new double[runs];
        for (int r = 0; r < runs; r++) {
            long a = System.nanoTime();
            f.run();
            long b = System.nanoTime();
            t[r] = (b - a) / 1e6;
        }
        Arrays.sort(t);
        double s = 0;
        for (double x : t) s += x;
        return new double[]{t[0], t[runs / 2], s / runs};
    }

    public static void main(String[] args) {
        int W = envi("BENCH_WIDTH", 5), D = envi("BENCH_DEPTH", 6),
            WARM = envi("BENCH_WARMUP", 3), RUNS = envi("BENCH_RUNS", 21),
            GP = envi("BENCH_GETPATH_ITERS", 2000);

        Object tree = build(W, D, 0);
        long nodes = nodecount(W, D);
        Object treeA = build(W, D, 1);
        Object treeB = build(W, D, 2);
        StringBuilder pb = new StringBuilder();
        for (int i = 0; i < D; i++) { if (i > 0) pb.append('.'); pb.append("k0"); }
        String path = pb.toString();

        Struct.WalkApply cb = (key, val, parent, p) -> { sink += p.size(); return val; };

        String[] names = {"clone", "walk", "merge", "stringify", "getpath"};
        long[] ucs = {nodes, nodes, nodes, nodes, GP};
        Fn[] fns = new Fn[]{
            () -> { if (Struct.clone(tree) != null) sink++; },
            () -> Struct.walk(tree, cb),
            () -> {
                List<Object> ml = new ArrayList<>();
                ml.add(treeA); ml.add(treeB);
                if (Struct.merge(ml) != null) sink++;
            },
            () -> { sink += Struct.stringify(tree).length(); },
            () -> {
                long s = 0;
                for (int i = 0; i < GP; i++) if (Struct.getpath(tree, path) != null) s++;
                sink += s;
            },
        };

        StringBuilder ops = new StringBuilder();
        for (int i = 0; i < names.length; i++) {
            double[] m = measure(WARM, RUNS, fns[i]);
            if (i > 0) ops.append(',');
            ops.append(String.format(
                "{\"op\":\"%s\",\"runs\":%d,\"unit_count\":%d,"
                + "\"min_ms\":%s,\"median_ms\":%s,\"mean_ms\":%s}",
                names[i], RUNS, ucs[i], m[0], m[1], m[2]));
        }

        System.err.println("java: sink=" + sink);
        System.out.println(String.format(
            "{\"lang\":\"java\",\"runtime\":\"java %s\",\"nodes\":%d,"
            + "\"params\":{\"width\":%d,\"depth\":%d,\"warmup\":%d,\"runs\":%d,"
            + "\"getpath_iters\":%d},\"ops\":[%s]}",
            System.getProperty("java.version"), nodes, W, D, WARM, RUNS, GP,
            ops.toString()));
    }
}

File: kotlin/bench/bench.kt
// Performance bench for the Kotlin port. Emits one JSON line per
// build/bench/README.md; diagnostics go to stderr.
import voxgig.struct.Struct

var sink = 0L

fun envi(k: String, d: Int): Int = System.getenv(k)?.toIntOrNull() ?: d

fun build(w: Int, d: Int, leaf: Int): Any {
    if (d == 0) return leaf
    val m = LinkedHashMap<String, Any?>()
    for (i in 0 until w) m["k$i"] = build(w, d - 1, leaf)
    return m
}

fun nodecount(w: Int, d: Int): Long {
    var n = 0L; var p = 1L
    for (i in 0..d) { n += p; p *= w }
    return n
}

fun measure(warm: Int, runs: Int, f: () -> Unit): DoubleArray {
    for (i in 0 until warm) f()
    val t = DoubleArray(runs)
    for (r in 0 until runs) { val a = System.nanoTime(); f(); t[r] = (System.nanoTime() - a) / 1e6 }
    t.sort()
    var s = 0.0; for (x in t) s += x
    return doubleArrayOf(t[0], t[runs / 2], s / runs)
}

fun main() {
    val W = envi("BENCH_WIDTH", 5); val D = envi("BENCH_DEPTH", 6)
    val WARM = envi("BENCH_WARMUP", 3); val RUNS = envi("BENCH_RUNS", 21)
    val GP = envi("BENCH_GETPATH_ITERS", 2000)

    val tree = build(W, D, 0)
    val nodes = nodecount(W, D)
    val treeA = build(W, D, 1)
    val treeB = build(W, D, 2)
    val path = (0 until D).joinToString(".") { "k0" }
    val cb = Struct.WalkApply { _, v, _, p -> sink += p.size; v }

    val names = arrayOf("clone", "walk", "merge", "stringify", "getpath")
    val ucs = longArrayOf(nodes, nodes, nodes, nodes, GP.toLong())
    val fns = arrayOf<() -> Unit>(
        { if (Struct.clone(tree) != null) sink++ },
        { Struct.walk(tree, cb) },
        { if (Struct.merge(listOf(treeA, treeB)) != null) sink++ },
        { sink += Struct.stringify(tree).length },
        { var a = 0L; for (i in 0 until GP) if (Struct.getpath(tree, path) != null) a++; sink += a },
    )

    val sb = StringBuilder()
    for (i in names.indices) {
        val m = measure(WARM, RUNS, fns[i])
        if (i > 0) sb.append(',')
        sb.append("{\"op\":\"${names[i]}\",\"runs\":$RUNS,\"unit_count\":${ucs[i]},\"min_ms\":${m[0]},\"median_ms\":${m[1]},\"mean_ms\":${m[2]}}")
    }

    System.err.println("kotlin: sink=$sink")
    println("{\"lang\":\"kotlin\",\"runtime\":\"kotlin/jvm ${System.getProperty("java.version")}\",\"nodes\":$nodes,\"params\":{\"width\":$W,\"depth\":$D,\"warmup\":$WARM,\"runs\":$RUNS,\"getpath_iters\":$GP},\"ops\":[$sb]}")
}
